logrotate: add design and api for logrotate (#37)
ceph/ceph-csi-operator: docs/design/logrotate.md

Reviewer: I don't think we need a separate MD for this. We do not do it for any other feature of the operator. Can we please merge this with an existing doc?

Reviewer 2: @parth-gr please update the existing doc as its a new field inside the current CR's?

Author: I think I can update the existing doc CRs, but on adding new features good to add the design doc flow, if someone is just interested in reading a specific logrotate feature or vice versa

Reviewer: @parth-gr Please keep the text here aligned with the code changes above

Author: Done

# Ceph CSI Log Rotate Design Document

Log Rotate is the ability to rotate the logs by controlling the size of log files.
With csi it will be includes to all the csi pods which includes csi daemons and provisoners.
Using logrotate will create a log file at the hostpath and enable the logrotate sidecar container.
Logrotate sidecar container will make sure that log files are rotated as per the inputs.

So logrotate will be enabled by,

`CephCSIOperatorConfig CRD`:
```yaml
spec:
    provisioner:
        privileged: true
```

Logroate configuration, 

`CephCSIOperatorConfig CRD`:

```yaml
kind: CephCSIOperatorConfig 
apiVersion: csi.ceph.io/v1alpha1
….
spec: 
logLevel: 1 
driverSpecDefaults: 
logging: 
    logLevel: 5
    # one of: hourly, daily, weekly
    Periodicity: daily

>     # SUFFIX may be 'M' or 'G' 

Reviewer 2: why we need this comment?

Author: Just for user clarification, if it is to easy to understand I can remove it

    MaxLogSize: 500M 
    MaxFiles: 5

>     logHostPath: /var/lib/rook 

Reviewer 2: use some other simple folder name like `/var/lib/log`

```

Similar settings will be overrided by `CephCSIDriver CRD`:

```yaml
kind: CephCSIDriver 
apiVersion: csi.ceph.io/v1alpha1 
metadata: 
name: "<prefix>.<driver_type>.csi.ceph.com" 
namespace:     <operator-namespace> 
spec: 
logLevel: 1 
driverSpecDefaults: 

> logging: 

Reviewer 2: it should be `log`

>     driverSpecDefaults: 

Author: How the `CephCSIDriver` is planned to override the defaults, from the current API I see I have to use the same `driverSpecDefaults` 

Or it be overridden under  `plugin` or `provisioner` heads?

    logLevel: 5
    # one of: hourly, daily, weekly
    Periodicity: daily
    # SUFFIX may be 'M' or 'G' 
    MaxLogSize: 500M 
    MaxFiles: 5
    logHostPath: /var/lib/rook 
```

> Logrotate sidecar container Cpu and memory usage can configured by,

Reviewer 2: `Cpu` to `cpu`

`CephCSIOperatorConfig CRD`:
```yaml
spec:
    provisioner:

>         LogRotate:

Reviewer 2: it should be `logRotator`

# Ceph CSI Log Rotate Design Document

> Log Rotate is the ability to rotate the logs by controlling the size of log files.
> With csi it will be includes to all the csi pods which includes csi daemons and provisoners.
> Using logrotate will create a log file at the hostpath and enable the logrotator sidecar container.
> Logrotator sidecar container will make sure that log files are rotated as per the inputs.

Reviewer 2: ```suggestion
Log rotation involves managing log files by controlling their size. This feature will be added to the CephCSI and csi-addons containers in the CephCSI node-plugin and controller-plugin pods. When logrotate is used, it will generate a log file at the host path and run the logrotator sidecar container in CephCSI pods. The logrotator sidecar container will then ensure that log files are rotated based on predefined settings.
```

Author: Thanks it look much clear

Logroate configuration, 

`CephCSIOperatorConfig CRD`:

```yaml
kind: CephCSIOperatorConfig 
apiVersion: csi.ceph.io/v1alpha1
….
spec: 
    logLevel: 1 
    driverSpecDefaults: 
        log:
            logLevel: 5
            logRotator:
                # one of: hourly, daily, weekly
                Periodicity: daily

>                 # SUFFIX may be 'M' or 'G' 

Reviewer 2: comments can be removed which are not related to this design like this one.

                MaxLogSize: 500M 
                MaxFiles: 5

>                 logHostPath: /var/lib/log 

Reviewer 3: `/var/lib/log` does not seem like a valid example. In the rook scenario, this would be set to `/var/lib/rook`, right? What path will be the default if not specified? Let's set this example to that default path.
```suggestion
                logHostPath: /var/lib/cephcsi
```

Author: DOne

>                 # one of: hourly, daily, weekly
>                 Periodicity: daily
>                 # SUFFIX may be 'M' or 'G' 
>                 MaxLogSize: 500M 
>                 MaxFiles: 5
>                 logHostPath: /var/lib/cephcsi 

Reviewer 2: ```suggestion
                # one of: hourly, daily, weekly
                periodicity: daily
                maxLogSize: 500M 
                maxFiles: 5
                logHostPath: /var/lib/cephcsi
```

```

Similar settings will be overrided by `CephCSIDriver CRD`:

```yaml
kind: CephCSIDriver 
apiVersion: csi.ceph.io/v1alpha1 

> metadata: 
> name: "<prefix>.<driver_type>.csi.ceph.com" 
> namespace:     <operator-namespace> 

Reviewer 2: ```suggestion
metadata: 
  name: "<prefix>.<driver_type>.csi.ceph.com" 
  namespace: <operator-namespace> 
```

spec: 
    logLevel: 1 
    driverSpecDefaults: 
        log: 
            logLevel: 5
            logRotator:

>                 # one of: hourly, daily, weekly
>                 Periodicity: daily
>                 # SUFFIX may be 'M' or 'G' 
>                 MaxLogSize: 500M 
>                 MaxFiles: 5
>                 logHostPath: /var/lib/rook 

Reviewer 2: ```suggestion
                # one of: hourly, daily, weekly
                periodicity: daily
                maxLogSize: 500M 
                maxFiles: 5
                logHostPath: /var/lib/cephcsi
```
